### ISSUE TITLE
Frame up

### ISSUE THREAD

**Issue by hosseinmoein** [fairtide/DataFrame#1]
I like what you did there for C++ data frame.
I also have a C++ data frame at https://github.com/hosseinmoein/DataFrame.
I would like to collaborate and compare notes.
Best,
Hossein Moein

**Comment by zhouyan** [comment]
Thanks for the interest. I had a brief look of your project and quite like some features of it, for example the async API etc. I think the two are based on some different design goals, which makes them complementary in some ways

Mine is designed around Apache Arrow which is great for exchange data between languages (C++, R, Julia, Python, Java etc) and as a column data format in general.  But I think its interface in C++ is not great to say the least. No one want to write a visitor every time they need to do something about the data. The goals of my project start with providing (relatively easy) zero-copy, type safe access to data already in Arrow format and for its future direction a lot of things are still in design phase.

**Comment by hosseinmoein** [comment]
Very interesting,
I had a chat with Wes before I started my project, about arrows (he and I worked together at AQR). After my chat and looking at arrow, I understood that arrow was meant to be “under the hood” for Pandas, not a C++ asset.
So I deliberately stayed away from that. I wanted something exclusively for C++ with no consideration to serve Python or R. Not that I do not like Python or R, but because they already have a reach environment. 
My design goals were also minimum copy, continuous memory space and complete type heterogeneity.

**Comment by zhouyan** [comment]
My opinion is that different languages have different strengths and use case, especially for statistical computing (my background was in computational statistics). For example, here is what I had in mind when I started the  project: I want to perform large scale simulation in C++ because the scale of the simulation demands the best performance, use R to do statistical analysis of the results because it has all the necessary tools and at that stage of analysis performance is no longer an issue and I don’t want to write a couple hundred lines of code in C++ to do something I can do in R with a simple library function call, and on top of that I want to use Python to manage my data because it takes a fraction of code to retrieve my data from a remote database in Python than doing it in C++ and use it for other workflow purposes. Now the only problem remains is how do I move my data around all three of them with minimum effort and good performance. And my data is not just numerical data (for which I used to use HDF5), but also categories, date times, etc. So far Arrow is the best choice I have seen.

In summary my goal is more about to get data in and out in C++ as fast as possible (fast meaning both in runtime performance and developing effort). By providing easy access, we can feed data into the C++ functions we want to use (computational kernels or a library such as Eigen) with minimal effort and return results back to a high level language with little effort.
